lib/active_merchant/billing/gateways/world_net_v2/token.rb

```diff
@@ -0,0 +1,75 @@
+require 'active_merchant/billing/base'
+require 'active_merchant/posts_data'
+require 'json'
+
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    module WorldNetV2 #:nodoc:
+      class Token
+        include PostsData
+
+        TerminalMissingError = Class.new(StandardError)
+
+        @lock = Mutex.new
+        class_attribute :merchant_api_key, instance_writer: false
+        class_attribute :token_data, instance_writer: false
+        class_attribute :token_expires_at, instance_writer: false
+
+        def self.fetch(options = {})
+          new(options).fetch
+        end
+
+        def self.lock
+          @lock
+        end
+
+        def initialize(options = {})
+          @options = options
+          
+          self.class.merchant_api_key = options[:merchant_api_key] if !merchant_api_key && options[:merchant_api_key]
+        end
+
+        def fetch
+          lock.synchronize { fetch_new! if token_expired? || token_data.nil? }
+          token_data
+        end
+
+        private
+
+        attr_reader :options
+
+        def fetch_new!
+          response = JSON.parse(ssl_get(url, { Authorization: "Basic #{merchant_api_key}"}))
+
+          terminal = response['allowedTerminals'][0]
+          raise TerminalMissingError.new if terminal.nil?
+
+          self.class.token_data = { token: response['token'], terminal: terminal }
+          self.class.token_expires_at = Time.now + Integer(response['expiresIn']) * 3600
+        end
+
+        def url
+          if test?
+            'https://testpayments.worldnettps.com/merchant/api/v1/account/authenticate'
+          else
+            raise
```

Comment:
Does this mean we have some more work to do in order to support prod?

Reply:
Correct, still waiting on production URLs since I think they might be merchant specific? I guess it depends on what Worldnet comes back with, though the existing deprecated XML API gateway integration uses the `payments` subdomain instead of the `testpayments` subdomain, so maybe it's a simple configuration.